torchx/lib/torchx/backend.ex

```diff
@@ -0,0 +1,274 @@
+      [:logical_and, :logical_or, :logical_xor] ++
+      [:outer]
+
+  for op <- binary_ops do
+    @impl true
+    def unquote(op)(out, left, right) do
+      NIF.unquote(op)(to_ref(left), to_ref(right)) |> from_ref(out)
+    end
+  end
+
+  unary_ops =
+    Enum.map(Nx.Shared.unary_math_funs(), &elem(&1, 0)) ++
+      [:abs, :bitwise_not, :ceil, :floor, :negate, :round, :sign]
+
+  # [:count_leading_zeros, :population_count]
+
+  for op <- unary_ops do
+    if {op, 1} in NIF.__info__(:functions) do
+      @impl true
+      def unquote(op)(out, tensor) do
+        NIF.unquote(op)(to_ref(tensor)) |> from_ref(out)
+      end
+    end
+  end
+
+  @impl true
+  def dot(
+        out,
+        %T{} = left,
+        _axes1,
+        %T{} = right,
+        _axes2
+      ) do
+    NIF.dot(to_ref(left), to_ref(right)) |> from_ref(out)
+  end
+
+  @impl true
+  def cholesky(%T{} = out, %T{} = t) do
+    NIF.cholesky(to_ref(t)) |> from_ref(out)
+  end
+
+  @impl true
+  def qr(
+        {q_holder, r_holder},
+        tensor,
+        opts
+      ),
+      do: NIF.qr(to_ref(tensor), opts[:mode] == :reduced) |> from_pair_ref({q_holder, r_holder})
+
+  @big_tensor_threshold_bytes 10_000_000
+
+  @impl true
+  def inspect(%T{type: {_, elem_size}} = tensor, inspect_opts) do
+    alias Inspect.Algebra, as: IA
+
+    limit = if(inspect_opts.limit == :infinity, do: nil, else: inspect_opts.limit + 1)
+
+    result =
+      if on_cpu?(tensor) do
+        byte_size = nbytes(tensor)
+        byte_limit = limit && limit * div(elem_size, 8)
+
+        if min(byte_limit, byte_size) > @big_tensor_threshold_bytes do
+          "Torchx tensor is too large to inspect. Explicitly transfer the tensor by calling Nx.backend_transfer/1"
+        else
+          binary = to_blob(tensor, limit)
+          Nx.Backend.inspect(tensor, binary, inspect_opts)
+        end
+      else
+        "Tensors on the GPU cannot be inspected. Explicitly transfer the tensor by calling Nx.backend_transfer/1"
+      end
+
+    IA.concat(["Torchx.Backend(#{device(tensor)})", IA.line(), result])
+  end
+
+  ## Helpers
+
+  defp unwrap!({:ok, result}), do: result
+
+  defp unwrap!({:error, error}),
+    do: raise(RuntimeError, "Torchx: " <> List.to_string(error))
+
+  defp from_ref(maybe_ref, t), do: maybe_ref |> unwrap!() |> to_tensor(t)
+
+  defp from_pair_ref(maybe_ref, {t1, t2}) do
+    {left, right} = unwrap!(maybe_ref)
+    {to_tensor(left, t1), to_tensor(right, t2)}
+  end
+
+  defp from_list_ref(maybe_ref, t),
+    do:
+      maybe_ref
+      |> unwrap!()
+      |> Enum.map(&to_tensor(&1, t))
+
+  defp to_ref(%T{data: %TB{ref: ref}}), do: ref
+
+  defp to_ref(%T{} = tensor),
+    do: Nx.backend_transfer(tensor, TB) |> to_ref()
+
+  # Update out tensor type here to mark the cases where our type policy mismatches with the libtorch's one.
+  defp to_tensor(ref, %T{} = t),
+    do: %T{t | data: %__MODULE__{ref: ref}, type: from_torch_type(unwrap!(NIF.type(ref)))}
```

Comment:
I think instead we want to do something like this:

```suggestion
  defp to_tensor(ref, %T{type: type} = t) do
    current_type = ref |> NIF.type() |> unwrap!() |> from_torch_type()

    if current != type do
      raise "type mismatch in Torchx: expected #{inspect(type)}, got: #{inspect(current_type)}. " <>
              "Please report this bug"
    end
    
    %{t | data: %__MODULE__{ref: ref}}
  end
```